File: ts/src/canary_validate.ts
import { existsSync, mkdirSync, readdirSync, readFileSync, writeFileSync } from "node:fs";
import { dirname, join, relative, resolve } from "node:path";
import type { ProviderName, RunKind, RunMode } from "./types";

type Args = {
  baselineResults: string[];
  candidateResults: string[];
  baselineResultsDir?: string;
  candidateResultsDir?: string;
  output?: string;
  format: "json" | "markdown";
  minReductionPct: number;
  maxPassDrop: number;
  maxWallRatio: number;
};

type BenchResultRow = {
  task_id: string;
  passed?: boolean;
  elapsed_seconds?: number;
  estimated_cost_usd?: number;
  resource_observation?: unknown;
  effective_resources?: unknown;
};

type BenchFile = {
  provider: ProviderName;
  mode?: RunMode;
  kind?: RunKind;
  task_count?: number;
  passed?: number;
  estimated_cost_usd?: number;
  results: BenchResultRow[];
};

type LoadedRun = {
  path: string;
  data: BenchFile;
};

type RunValidation = {
  key: string;
  baseline_path: string;
  candidate_path: string;
  provider: ProviderName;
  mode?: RunMode;
  kind?: RunKind;
  matched_tasks: number;
  baseline_passed: number;
  candidate_passed: number;
  pass_drop: number;
  baseline_elapsed_seconds: number;
  candidate_elapsed_seconds: number;
  wall_ratio?: number;
  baseline_cost_usd: number;
  candidate_cost_usd: number;
  reduction_pct?: number;
  candidate_observations: number;
  passed: boolean;
  failures: string[];
};

function parseArgs(argv: string[]): Args {
  const values = new Map<string, string>();
  for (let index = 0; index < argv.length; index += 2) {
    values.set(argv[index], argv[index + 1]);
  }
  return {
    baselineResults: parseList(values.get("--baseline-results") ?? values.get("--baseline")),
    candidateResults: parseList(values.get("--candidate-results") ?? values.get("--candidate")),
    baselineResultsDir: values.get("--baseline-results-dir"),
    candidateResultsDir: values.get("--candidate-results-dir"),
    output: values.get("--output"),
    format: parseFormat(values.get("--format") ?? "markdown"),
    minReductionPct: Number.parseFloat(values.get("--min-reduction-pct") ?? "20"),
    maxPassDrop: Number.parseInt(values.get("--max-pass-drop") ?? "0", 10),
    maxWallRatio: Number.parseFloat(values.get("--max-wall-ratio") ?? "1.2")
  };
}

function parseList(value: string | undefined): string[] {
  return (value ?? "")
    .split(",")
    .map((item) => item.trim())
    .filter(Boolean);
}

function parseFormat(value: string): Args["format"] {
  if (value === "json" || value === "markdown") {
    return value;
  }
  throw new Error(`Unsupported format: ${value}`);
}

function resultPaths(paths: string[], dir: string | undefined): string[] {
  const all = [...paths];
  if (dir) {
    all.push(...readdirSync(dir).filter((file) => file.endsWith(".json")).map((file) => join(dir, file)));
  }
  if (all.length === 0) {
    throw new Error("No results provided.");
  }
  return all.map((path) => resolve(path));
}

async function loadRuns(paths: string[]): Promise<LoadedRun[]> {
  return await Promise.all(paths.map(loadRun));
}

async function loadRun(path: string): Promise<LoadedRun> {
  if (!existsSync(path)) {
    throw new Error(`Result input not found: ${path}`);
  }
  const data = (await Bun.file(path).json()) as BenchFile;
  if (!data.provider || !Array.isArray(data.results)) {
    throw new Error(`Unsupported benchmark result file: ${path}`);
  }
  return { path, data };
}

function runKey(data: BenchFile): string {
  return [data.provider, data.mode ?? "unknown-mode", data.kind ?? "unknown-kind"].join("|");
}

function validateRun(candidate: LoadedRun, baselines: Map<string, LoadedRun>, args: Args): RunValidation {
  const key = runKey(candidate.data);
  const baseline = baselines.get(key) ?? baselines.get([candidate.data.provider, "unknown-mode", "unknown-kind"].join("|"));
  if (!baseline) {
    throw new Error(`No baseline result found for candidate ${candidate.path} (${key})`);
  }
  const baselineRows = new Map(baseline.data.results.map((row) => [row.task_id, row]));
  const matched = candidate.data.results.flatMap((candidateRow) => {
    const baselineRow = baselineRows.get(candidateRow.task_id);
    return baselineRow ? [{ baselineRow, candidateRow }] : [];
  });
  if (matched.length === 0) {
    throw new Error(`No overlapping task IDs between ${baseline.path} and ${candidate.path}`);
  }
  const baselineCost = sum(matched.map(({ baselineRow }) => rowCost(baselineRow, baseline.data)));
  const candidateCost = sum(matched.map(({ candidateRow }) => rowCost(candidateRow, candidate.data)));
  const baselineElapsed = sum(matched.map(({ baselineRow }) => finiteNumber(baselineRow.elapsed_seconds) ?? 0));
  const candidateElapsed = sum(matched.map(({ candidateRow }) => finiteNumber(candidateRow.elapsed_seconds) ?? 0));
  const baselinePassed = matched.filter(({ baselineRow }) => baselineRow.passed === true).length;
  const candidatePassed = matched.filter(({ candidateRow }) => candidateRow.passed === true).length;
  const passDrop = baselinePassed - candidatePassed;
  const reductionPct = pctReduction(baselineCost, candidateCost);
  const wallRatio = baselineElapsed > 0 ? candidateElapsed / baselineElapsed : undefined;
  const failures = [
    ...(reductionPct === undefined || reductionPct < args.minReductionPct
      ? [`cost reduction ${formatPct(reductionPct)} below ${args.minReductionPct.toFixed(1)}%`]
      : []),
    ...(passDrop > args.maxPassDrop ? [`pass drop ${passDrop} above ${args.maxPassDrop}`] : []),
    ...(wallRatio !== undefined && wallRatio > args.maxWallRatio
      ? [`wall ratio ${wallRatio.toFixed(2)} above ${args.maxWallRatio.toFixed(2)}`]
      : [])
  ];
  return {
    key,
    baseline_path: baseline.path,
    candidate_path: candidate.path,
    provider: candidate.data.provider,
    ...(candidate.data.mode ? { mode: candidate.data.mode } : {}),
    ...(candidate.data.kind ? { kind: candidate.data.kind } : {}),
    matched_tasks: matched.length,
    baseline_passed: baselinePassed,
    candidate_passed: candidatePassed,
    pass_drop: passDrop,
    baseline_elapsed_seconds: baselineElapsed,
    candidate_elapsed_seconds: candidateElapsed,
    ...(wallRatio === undefined ? {} : { wall_ratio: wallRatio }),
    baseline_cost_usd: baselineCost,
    candidate_cost_usd: candidateCost,
    ...(reductionPct === undefined ? {} : { reduction_pct: reductionPct }),
    candidate_observations: matched.filter(({ candidateRow }) => candidateRow.resource_observation !== undefined).length,
    passed: failures.length === 0,
    failures
  };
}

function rowCost(row: BenchResultRow, run: BenchFile): number {
  const direct = finiteNumber(row.estimated_cost_usd);
  if (direct !== undefined) {
    return direct;
  }
  const runCost = finiteNumber(run.estimated_cost_usd);
  if (runCost === undefined) {
    return 0;
  }
  const elapsed = finiteNumber(row.elapsed_seconds) ?? 0;
  const totalElapsed = sum(run.results.map((item) => finiteNumber(item.elapsed_seconds) ?? 0));
  return totalElapsed > 0 ? runCost * (elapsed / totalElapsed) : 0;
}

function finiteNumber(value: unknown): number | undefined {
  return typeof value === "number" && Number.isFinite(value) ? value : undefined;
}

function pctReduction(baseline: number, candidate: number): number | undefined {
  return baseline > 0 ? ((baseline - candidate) / baseline) * 100 : undefined;
}

function sum(values: number[]): number {
  return values.reduce((total, value) => total + value, 0);
}

function markdownReport(validations: RunValidation[], args: Args): string {
  const baselineCost = sum(validations.map((item) => item.baseline_cost_usd));
  const candidateCost = sum(validations.map((item) => item.candidate_cost_usd));
  const baselineElapsed = sum(validations.map((item) => item.baseline_elapsed_seconds));
  const candidateElapsed = sum(validations.map((item) => item.candidate_elapsed_seconds));
  const baselinePassed = sum(validations.map((item) => item.baseline_passed));
  const candidatePassed = sum(validations.map((item) => item.candidate_passed));
  const totalReduction = pctReduction(baselineCost, candidateCost);
  return [
    "# Canary Cost Validation",
    "",
    `Generated: ${new Date().toISOString()}`,
    "",
    `Minimum reduction: ${args.minReductionPct.toFixed(1)}%`,
    `Maximum pass drop: ${args.maxPassDrop}`,
    `Maximum wall ratio: ${args.maxWallRatio.toFixed(2)}`,
    "",
    "## Summary",
    "",
    "baseline cost | candidate cost | reduction | baseline passed | candidate passed | wall ratio | status",
    "---: | ---: | ---: | ---: | ---: | ---: | ---",
    [
      formatMoney(baselineCost),
      formatMoney(candidateCost),
      formatPct(totalReduction),
      String(baselinePassed),
      String(candidatePassed),
      baselineElapsed > 0 ? (candidateElapsed / baselineElapsed).toFixed(2) : "-",
      validations.every((item) => item.passed) ? "pass" : "fail"
    ].join(" | "),
    "",
    "## Runs",
    "",
    "provider | mode | tasks | cost reduction | pass drop | wall ratio | observations | status | failures",
    "--- | --- | ---: | ---: | ---: | ---: | ---: | --- | ---",
    ...validations.map((item) =>
      [
        item.provider,
        item.mode ?? "-",
        String(item.matched_tasks),
        formatPct(item.reduction_pct),
        String(item.pass_drop),
        item.wall_ratio === undefined ? "-" : item.wall_ratio.toFixed(2),
        String(item.candidate_observations),
        item.passed ? "pass" : "fail",
        item.failures.join("; ") || "-"
      ].join(" | ")
    ),
    "",
    "## Inputs",
    "",
    ...validations.flatMap((item) => [
      `- ${item.provider}${item.mode ? ` ${item.mode}` : ""}: baseline \`${displayPath(item.baseline_path)}\`, candidate \`${displayPath(item.candidate_path)}\``
    ]),
    ""
  ].join("\n");
}

function formatMoney(value: number | undefined): string {
  return value === undefined || Number.isNaN(value) ? "-" : `$${value.toFixed(4)}`;
}

function formatPct(value: number | undefined): string {
  return value === undefined || Number.isNaN(value) ? "-" : `${value.toFixed(1)}%`;
}

function displayPath(path: string): string {
  const rel = relative(process.cwd(), path);
  return rel.startsWith("..") ? path : rel;
}

function writeOutput(path: string | undefined, content: string): void {
  if (!path) {
    console.log(content);
    return;
  }
  mkdirSync(dirname(path), { recursive: true });
  writeFileSync(path, content);
  console.log(`wrote ${path}`);
}

async function main(): Promise<void> {
  const args = parseArgs(Bun.argv.slice(2));
  const [baselineRuns, candidateRuns] = await Promise.all([
    loadRuns(resultPaths(args.baselineResults, args.baselineResultsDir)),
    loadRuns(resultPaths(args.candidateResults, args.candidateResultsDir))
  ]);
  const baselines = new Map(baselineRuns.map((run) => [runKey(run.data), run]));
  const validations = candidateRuns
    .map((run) => validateRun(run, baselines, args))
    .sort((left, right) => left.provider.localeCompare(right.provider) || (left.mode ?? "").localeCompare(right.mode ?? ""));
  const payload = {
    schema_version: 1,
    generated_at: new Date().toISOString(),
    min_reduction_pct: args.minReductionPct,
    max_pass_drop: args.maxPassDrop,
    max_wall_ratio: args.maxWallRatio,
    passed: validations.every((item) => item.passed),
    total_baseline_cost_usd: sum(validations.map((item) => item.baseline_cost_usd)),
    total_candidate_cost_usd: sum(validations.map((item) => item.candidate_cost_usd)),
    total_reduction_pct: pctReduction(
      sum(validations.map((item) => item.baseline_cost_usd)),
      sum(validations.map((item) => item.candidate_cost_usd))
    ),
    validations
  };
  const content = args.format === "json" ? `${JSON.stringify(payload, null, 2)}\n` : markdownReport(validations, args);
  writeOutput(args.output, content);
  if (!payload.passed) {
    process.exitCode = 1;
  }
}

if (import.meta.main) {
  await main();
}

File: py/code_sandbox_bench/canary_validate.py
import argparse
import json
import math
from datetime import datetime, timezone
from pathlib import Path
from typing import Any


def iso_now() -> str:
    return datetime.now(timezone.utc).isoformat().replace("+00:00", "Z")


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser()
    parser.add_argument("--baseline-results", "--baseline", default="")
    parser.add_argument("--candidate-results", "--candidate", default="")
    parser.add_argument("--baseline-results-dir", type=Path)
    parser.add_argument("--candidate-results-dir", type=Path)
    parser.add_argument("--output", type=Path)
    parser.add_argument("--format", choices=["json", "markdown"], default="markdown")
    parser.add_argument("--min-reduction-pct", type=float, default=20.0)
    parser.add_argument("--max-pass-drop", type=int, default=0)
    parser.add_argument("--max-wall-ratio", type=float, default=1.2)
    return parser.parse_args()


def result_paths(paths_text: str, directory: Path | None) -> list[Path]:
    paths = [Path(item) for item in paths_text.split(",") if item.strip()]
    if directory:
        paths.extend(path for path in directory.iterdir() if path.suffix == ".json")
    if not paths:
        raise ValueError("No results provided.")
    return [path.resolve() for path in paths]


def load_run(path: Path) -> dict[str, Any]:
    if not path.exists():
        raise FileNotFoundError(f"Result input not found: {path}")
    data = json.loads(path.read_text(encoding="utf-8"))
    if not isinstance(data, dict) or not data.get("provider") or not isinstance(data.get("results"), list):
        raise ValueError(f"Unsupported benchmark result file: {path}")
    return {"path": str(path), "data": data}


def run_key(data: dict[str, Any]) -> str:
    return "|".join([str(data.get("provider")), str(data.get("mode") or "unknown-mode"), str(data.get("kind") or "unknown-kind")])


def validate_run(candidate: dict[str, Any], baselines: dict[str, dict[str, Any]], args: argparse.Namespace) -> dict[str, Any]:
    data = candidate["data"]
    key = run_key(data)
    baseline = baselines.get(key) or baselines.get("|".join([str(data.get("provider")), "unknown-mode", "unknown-kind"]))
    if baseline is None:
        raise ValueError(f"No baseline result found for candidate {candidate['path']} ({key})")
    baseline_rows = {
        row["task_id"]: row
        for row in baseline["data"]["results"]
        if isinstance(row, dict) and isinstance(row.get("task_id"), str)
    }
    matched = [
        {"baseline": baseline_rows[row["task_id"]], "candidate": row}
        for row in data["results"]
        if isinstance(row, dict) and isinstance(row.get("task_id"), str) and row["task_id"] in baseline_rows
    ]
    if not matched:
        raise ValueError(f"No overlapping task IDs between {baseline['path']} and {candidate['path']}")
    baseline_cost = sum(row_cost(pair["baseline"], baseline["data"]) for pair in matched)
    candidate_cost = sum(row_cost(pair["candidate"], data) for pair in matched)
    baseline_elapsed = sum(finite_number(pair["baseline"].get("elapsed_seconds")) or 0 for pair in matched)
    candidate_elapsed = sum(finite_number(pair["candidate"].get("elapsed_seconds")) or 0 for pair in matched)
    baseline_passed = sum(1 for pair in matched if pair["baseline"].get("passed") is True)
    candidate_passed = sum(1 for pair in matched if pair["candidate"].get("passed") is True)
    pass_drop = baseline_passed - candidate_passed
    reduction = pct_reduction(baseline_cost, candidate_cost)
    wall_ratio = candidate_elapsed / baseline_elapsed if baseline_elapsed > 0 else None
    failures = []
    if reduction is None or reduction < args.min_reduction_pct:
        failures.append(f"cost reduction {format_pct(reduction)} below {args.min_reduction_pct:.1f}%")
    if pass_drop > args.max_pass_drop:
        failures.append(f"pass drop {pass_drop} above {args.max_pass_drop}")
    if wall_ratio is not None and wall_ratio > args.max_wall_ratio:
        failures.append(f"wall ratio {wall_ratio:.2f} above {args.max_wall_ratio:.2f}")
    return {
        "key": key,
        "baseline_path": baseline["path"],
        "candidate_path": candidate["path"],
        "provider": data["provider"],
        **({"mode": data["mode"]} if data.get("mode") else {}),
        **({"kind": data["kind"]} if data.get("kind") else {}),
        "matched_tasks": len(matched),
        "baseline_passed": baseline_passed,
        "candidate_passed": candidate_passed,
        "pass_drop": pass_drop,
        "baseline_elapsed_seconds": baseline_elapsed,
        "candidate_elapsed_seconds": candidate_elapsed,
        **({"wall_ratio": wall_ratio} if wall_ratio is not None else {}),
        "baseline_cost_usd": baseline_cost,
        "candidate_cost_usd": candidate_cost,
        **({"reduction_pct": reduction} if reduction is not None else {}),
        "candidate_observations": sum(1 for pair in matched if pair["candidate"].get("resource_observation") is not None),
        "passed": not failures,
        "failures": failures,
    }


def row_cost(row: dict[str, Any], run: dict[str, Any]) -> float:
    direct = finite_number(row.get("estimated_cost_usd"))
    if direct is not None:
        return direct
    run_cost = finite_number(run.get("estimated_cost_usd"))
    if run_cost is None:
        return 0.0
    elapsed = finite_number(row.get("elapsed_seconds")) or 0.0
    total_elapsed = sum(finite_number(item.get("elapsed_seconds")) or 0.0 for item in run.get("results", []) if isinstance(item, dict))
    return run_cost * (elapsed / total_elapsed) if total_elapsed > 0 else 0.0


def markdown_report(validations: list[dict[str, Any]], args: argparse.Namespace) -> str:
    baseline_cost = sum(float(item["baseline_cost_usd"]) for item in validations)
    candidate_cost = sum(float(item["candidate_cost_usd"]) for item in validations)
    baseline_elapsed = sum(float(item["baseline_elapsed_seconds"]) for item in validations)
    candidate_elapsed = sum(float(item["candidate_elapsed_seconds"]) for item in validations)
    baseline_passed = sum(int(item["baseline_passed"]) for item in validations)
    candidate_passed = sum(int(item["candidate_passed"]) for item in validations)
    lines = [
        "# Canary Cost Validation",
        "",
        f"Generated: {iso_now()}",
        "",
        f"Minimum reduction: {args.min_reduction_pct:.1f}%",
        f"Maximum pass drop: {args.max_pass_drop}",
        f"Maximum wall ratio: {args.max_wall_ratio:.2f}",
        "",
        "## Summary",
        "",
        "baseline cost | candidate cost | reduction | baseline passed | candidate passed | wall ratio | status",
        "---: | ---: | ---: | ---: | ---: | ---: | ---",
        " | ".join(
            [
                format_money(baseline_cost),
                format_money(candidate_cost),
                format_pct(pct_reduction(baseline_cost, candidate_cost)),
                str(baseline_passed),
                str(candidate_passed),
                f"{candidate_elapsed / baseline_elapsed:.2f}" if baseline_elapsed > 0 else "-",
                "pass" if all(item["passed"] for item in validations) else "fail",
            ]
        ),
        "",
        "## Runs",
        "",
        "provider | mode | tasks | cost reduction | pass drop | wall ratio | observations | status | failures",
        "--- | --- | ---: | ---: | ---: | ---: | ---: | --- | ---",
    ]
    for item in validations:
        lines.append(
            " | ".join(
                [
                    str(item["provider"]),
                    str(item.get("mode") or "-"),
                    str(item["matched_tasks"]),
                    format_pct(item.get("reduction_pct")),
                    str(item["pass_drop"]),
                    f"{item['wall_ratio']:.2f}" if finite_number(item.get("wall_ratio")) else "-",
                    str(item["candidate_observations"]),
                    "pass" if item["passed"] else "fail",
                    "; ".join(item["failures"]) or "-",
                ]
            )
        )
    return "\n".join(lines + [""])


def finite_number(value: Any) -> float | None:
    return float(value) if isinstance(value, (int, float)) and math.isfinite(value) else None


def pct_reduction(baseline: float, candidate: float) -> float | None:
    return ((baseline - candidate) / baseline) * 100 if baseline > 0 else None


def format_money(value: Any) -> str:
    return "-" if finite_number(value) is None else f"${float(value):.4f}"


def format_pct(value: Any) -> str:
    return "-" if finite_number(value) is None else f"{float(value):.1f}%"


def write_output(path: Path | None, content: str) -> None:
    if path is None:
        print(content)
        return
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(content, encoding="utf-8")
    print(f"wrote {path}")


def main() -> None:
    args = parse_args()
    baseline_runs = [load_run(path) for path in result_paths(args.baseline_results, args.baseline_results_dir)]
    candidate_runs = [load_run(path) for path in result_paths(args.candidate_results, args.candidate_results_dir)]
    baselines = {run_key(run["data"]): run for run in baseline_runs}
    validations = sorted(
        [validate_run(run, baselines, args) for run in candidate_runs],
        key=lambda item: (str(item["provider"]), str(item.get("mode") or "")),
    )
    total_baseline = sum(float(item["baseline_cost_usd"]) for item in validations)
    total_candidate = sum(float(item["candidate_cost_usd"]) for item in validations)
    payload = {
        "schema_version": 1,
        "generated_at": iso_now(),
        "min_reduction_pct": args.min_reduction_pct,
        "max_pass_drop": args.max_pass_drop,
        "max_wall_ratio": args.max_wall_ratio,
        "passed": all(item["passed"] for item in validations),
        "total_baseline_cost_usd": total_baseline,
        "total_candidate_cost_usd": total_candidate,
        "total_reduction_pct": pct_reduction(total_baseline, total_candidate),
        "validations": validations,
    }
    content = json.dumps(payload, indent=2) + "\n" if args.format == "json" else markdown_report(validations, args)
    write_output(args.output, content)
    if not payload["passed"]:
        raise SystemExit(1)


if __name__ == "__main__":
    main()
